lib/TPP/Transforms/LoopInsertion.cpp

```diff
@@ -0,0 +1,307 @@
+  if (isa<memref::SubViewOp>(nDefiningOp))
+    nShape =
+        dyn_cast<ShapedType>(nDefiningOp->getOperand(0).getType()).getShape();
+  else
+    nShape =
+        dyn_cast<ShapedType>(nDefiningOp->getResult(0).getType()).getShape();
+
+  ArrayRef<int64_t> kShape;
+  auto kDefiningOp = brgemmOp.getOperand(3).getDefiningOp();
+  if (isa<memref::SubViewOp>(kDefiningOp))
+    kShape =
+        dyn_cast<ShapedType>(kDefiningOp->getOperand(0).getType()).getShape();
+  else
+    kShape =
+        dyn_cast<ShapedType>(kDefiningOp->getResult(0).getType()).getShape();
+
+  if (mShape.size() != kShape.size())
+    return rewriter.notifyMatchFailure(
+        op, "require m tensor shape and k tensor shape to match");
+
+  SmallVector<unsigned> tileShapeM;
+  tileShapeM.push_back(mTileShape);
+  tileShapeM.push_back(mShape[0] / mTileShape);
+
+  SmallVector<unsigned> tileShapeN;
+  tileShapeN.push_back(nTileShape);
+  tileShapeN.push_back(nShape[0] / nTileShape);
+
+  // Validate the input tile sizes against the operand shapes
+  long multipleM = 1;
+  for (size_t i = 0; i < tileShapeM.size(); i++)
+    multipleM = multipleM * tileShapeM[i];
+
+  if (mShape[0] != multipleM)
+    return rewriter.notifyMatchFailure(
+        op, "require m tile shape to match tensor shape");
+
+  long multipleN = 1;
+  for (size_t i = 0; i < tileShapeN.size(); i++)
+    multipleN = multipleN * tileShapeN[i];
+
+  if (nShape[0] != multipleN)
+    return rewriter.notifyMatchFailure(
+        op, "require n tile shape to match tensor shape");
+
+  if ((multipleM * multipleN) != (kShape[0] * kShape[1]))
+    return rewriter.notifyMatchFailure(
+        op, "require k tile shape to match tensor shape");
+
+  int boundSize = tileShapeM.size() + tileShapeN.size();
+
+  // Set the new bounds of for loop
+  SmallVector<int64_t> lbs(boundSize, 0), steps(boundSize, 1);
+
+  SmallVector<int64_t> ubs(tileShapeM.begin(), tileShapeM.end());
+  ubs.append(tileShapeN.begin(), tileShapeN.end());
+
+  op.setStaticLowerBound(lbs);
+  op.setStaticUpperBound(ubs);
+  op.setStaticStep(steps);
+
+  // Add new induction var args to the for loop
+  int numArgs = op.getBody()->getArguments().size();
+
+  for (int i = 0; i < boundSize - numArgs; i++)
+    op.getBody()->addArgument(b.getIndexType(), op.getLoc());
+
+  SmallVector<int64_t> tileOffsets{
+      0, static_cast<int64_t>(tileShapeM.size()),
+      static_cast<int64_t>(tileShapeN.size() + tileShapeM.size())};
+  rewriter.setInsertionPointToStart(op.getBody());
+
+  SmallVector<ArrayRef<int64_t>> originalShapes{mShape, nShape, kShape};
+  SmallVector<SmallVector<SmallVector<unsigned>>> tilingVectors{
+      {tileShapeM}, {tileShapeN}, {tileShapeM, tileShapeN}};
+
+  // Replace old args with newly computed args
+  for (auto oper = op.getBody()->getOperations().begin();
+       oper != op.getBody()->getOperations().end(); oper++) {
+    int operandIndex = 0;
+    for (auto arg : oper->getOperands()) {
```

Comment:
This logic is very confusing. What are you trying to do here?

Reply:
I have added a comment to explain this.